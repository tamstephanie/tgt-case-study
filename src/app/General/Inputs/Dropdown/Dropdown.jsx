import React, {Component} from "react";
import PropTypes from "prop-types";
import _ from "lodash";
import {FormControl, InputLabel, MenuItem, Select, withStyles} from "@material-ui/core";

/**
 * Basic component that renders a Material UI Select dropdown
 */
class Dropdown extends Component {
    render() {
        let {disabled, emptyOption, classes, color, inputLabel, options, selectedValue} = this.props;
        return (
            <FormControl className={classes.dropdown} disabled={disabled} fullWidth color={color}>
                {!_.isEmpty(inputLabel) && (
                    <InputLabel>{inputLabel}</InputLabel>
                )}
                <Select
                    value={selectedValue}
                    onChange={this.props.onChange}
                    displayEmpty={!_.isEmpty(emptyOption) ? true : false}
                    variant="outlined"
                >
                    {!_.isEmpty(emptyOption) && (
                        <MenuItem className={classes.optionItem} key="empty" value="">{emptyOption}</MenuItem>
                    )}
                    {_.map(options, (option) => {
                        let label, value;
                        if (_.isString(option) || _.isNumber(option)) {
                            value = option;
                        } else if (_.isObject(option)) {
                            value = _.get(option, "value");
                            label = _.get(option, "label");
                        }

                        return (
                            <MenuItem className={classes.optionItem} key={value} value={value}>
                                {!_.isNil(label) ? label : value}
                            </MenuItem>
                        );
                    })}
                </Select>
            </FormControl>
        );
    }

}

Dropdown.propTypes = {
    /**
     * Function that controls the behavior of the dropdown 
     * @type {Function}
     */
    onChange: PropTypes.func.isRequired,
    /**
     * Any acceptable value for [MuiSelect]
     * @type {String | Number | Array}
     */
    selectedValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.array]),
    /**
     * Indicates if the dropdown should be disabled
     */
    disabled: PropTypes.bool,
    /**
     * The text to display if an empty option is allowed
     * @type {String}
     */
    emptyOption: PropTypes.string,
    /**
     * An input label to display
     * @type {String}
     */
    inputLabel: PropTypes.string,
    /**
     * Indicates if multiple values are allowed at once
     * @type {Boolean}
     */
    multi: PropTypes.bool,
    /**
     * List of dropdown options
     * @type {Number []| String[] | Object[]}
     */
    options: PropTypes.arrayOf(PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
        PropTypes.shape({
            value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
            label: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
        })
    ])),
    color: PropTypes.oneOf(["primary", "secondary"])
};

Dropdown.defaultProps = {
    disabled: false,
    emptyOption: "",
    inputLabel: "",
    multi: false,
    options: []
};

const style = (theme) => ({
    dropdown: {
        "& .MuiSelect-root": {
            fontSize: "1.15rem",
            width: "100%"
        }
    },
    optionItem: {
        fontSize: "1.15rem"
    }
});

export default withStyles(style)(Dropdown);