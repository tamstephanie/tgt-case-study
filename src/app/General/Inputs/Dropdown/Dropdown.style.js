export const DropdownStyle =  (theme) => ({
    dropdown: {
        "& .MuiSelect-root": {
            fontSize: "1.15rem",
            width: "100%"
        }
    },
    optionItem: {
        fontSize: "1.15rem"
    }
});