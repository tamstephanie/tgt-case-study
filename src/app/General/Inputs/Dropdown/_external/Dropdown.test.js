import React from "react";
import _ from "lodash";

import chai from "chai";
import {mount} from "enzyme";

import Dropdown from "../Dropdown";
import {OPTIONS_COMPLEX, OPTIONS_SIMPLE} from "./data.mock";

const COMPONENT_NAME = "Dropdown";

describe(COMPONENT_NAME, () => {
    it("displays a dropdown when given a simple list of strings", () => {
        let dropdown = mount(<Dropdown onChange={jest.fn()} options={OPTIONS_SIMPLE} />);

        let menuItems = dropdown.find("ForwardRef(MenuItem)");
        chai.expect(menuItems).to.have.lengthOf(12);
        _.forEach(menuItems, (menuItem) => {
            chai.expect(_.include(OPTIONS_SIMPLE, menuItem.text())).to.be.true;
        });
    })
});