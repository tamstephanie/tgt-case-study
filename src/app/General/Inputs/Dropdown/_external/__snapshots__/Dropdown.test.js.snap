// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders a dropdown with simple options 1`] = `
<div
  className="MuiFormControl-root Dropdown-dropdown-1 MuiFormControl-fullWidth"
>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
    onClick={[Function]}
  >
    <div
      aria-haspopup="listbox"
      className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      Select a fruit
    </div>
    <input
      aria-hidden={true}
      className="MuiSelect-nativeInput"
      onAnimationStart={[Function]}
      onChange={[Function]}
      required={false}
      tabIndex={-1}
      value=""
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
    <fieldset
      aria-hidden={true}
      className="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
      style={
        Object {
          "paddingLeft": 8,
        }
      }
    >
      <legend
        className="PrivateNotchedOutline-legend-4"
        style={
          Object {
            "width": 0.01,
          }
        }
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203;",
            }
          }
        />
      </legend>
    </fieldset>
  </div>
</div>
`;
